workflow/controller/operator.go

```diff
@@ -3074,76 +3091,70 @@ func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
+		return tmpl.RetryStrategy
+	}
+	return woc.execWf.Spec.RetryStrategy
+}
 
-	if woc.wf.Spec.WorkflowTemplateRef == nil {
-		if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
-			return nil, executionParameters, fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+func (woc *wfOperationCtx) setExecWorkflow() error {
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		err := woc.setStoredWFSpec()
+		if err != nil {
+			return err
 		}
-		// Set the WorkflowDefaults from Configmap
+		woc.execWf = &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec.DeepCopy()}
+		woc.volumes = woc.execWf.Spec.DeepCopy().Volumes
+	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
+		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+	} else {
 		err := woc.controller.setWorkflowDefaults(woc.wf)
 		if err != nil {
-			log.WithFields(log.Fields{"key": woc.wf.Name, "error": err}).Warn("Failed to apply default workflow values")
-			return nil, executionParameters, err
+			return err
 		}
-
-		tmplRef := &wfv1.WorkflowStep{Template: woc.wf.Spec.Entrypoint}
-		return tmplRef, executionParameters, nil
+		woc.volumes = woc.wf.Spec.DeepCopy().Volumes
 	}
+	return nil
+}
 
+func (woc *wfOperationCtx) setStoredWFSpec() error {
 	if woc.wf.Status.StoredWorkflowSpec == nil {
-		wftSpec, err := woc.fetchWorkflowSpec()
+		wfDefault := woc.controller.Config.WorkflowDefaults
+		wftHolder, err := woc.fetchWorkflowSpec()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		woc.wf.Status.StoredWorkflowSpec = wftSpec
+		// Join WFT and WfDefault metadata to Workflow metadata.
+		wfutil.JoinWorkflowMetaData(&woc.wf.ObjectMeta, wftHolder.GetWorkflowMetadata(), &wfDefault.ObjectMeta)
+
+		// Join WF spec WFT spec and WorkflowDefault spec
+		mergedWf, err := wfutil.JoinWorkflowSpec(&woc.wf.Spec, wftHolder.GetWorkflowSpec(), &wfDefault.Spec)
+		if err != nil {
+			return err
+		}
+		// Clear the WorkflowTemplateRef from merged Workflow
+		mergedWf.Spec.WorkflowTemplateRef = nil
+
+		woc.wf.Status.StoredWorkflowSpec = &mergedWf.Spec
 		woc.updated = true
 	} else if woc.controller.Config.WorkflowRestrictions.MustNotChangeSpec() {
-		// If the controller is in reference mode, ensure that the stored spec is identical to the reference spec at every operation
-		wftSpec, err := woc.fetchWorkflowSpec()
+		wftHolder, err := woc.fetchWorkflowSpec()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		if woc.wf.Status.StoredWorkflowSpec.String() != wftSpec.String() {
-			return nil, executionParameters, fmt.Errorf("workflowTemplateRef reference may not change during execution when the controller is in reference mode")
+		wfDefault := woc.controller.Config.WorkflowDefaults
+		mergedWf, err := wfutil.JoinWorkflowSpec(&woc.wf.Spec, wftHolder.GetWorkflowSpec(), &wfDefault.Spec)
```

Comment:
What if workflow defaults change during execution? That would trigger this failure. Is that something we want? Unsure because the cluster admin has access to workflow defaults, so we might not want to consider that a breaking change to the execution.

Reply:
Yes, We (Jesse & Alex) discussed this behavior. We need to fail it